### PR TITLE
[SYCL] Introduce new properties for virtual functions

### PR DESCRIPTION
**Pull request by AlexeySachkov**
See language specification for those properties in intel/llvm#10540

### CODE REVIEW

File: sycl/include/sycl/handler.hpp [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include <sycl/exception.hpp>
 #include <sycl/ext/intel/experimental/fp_control_kernel_properties.hpp>
 #include <sycl/ext/intel/experimental/kernel_execution_properties.hpp>
+#include <sycl/ext/intel/experimental/virtual_functions.hpp>
```
Comment (AlexeySachkov) [inline sycl/include/sycl/handler.hpp:31]: Note for reviewers: I've tried not to include this here, but a bigger refactoring is needed if we won't want to see includes like this in `handler.hpp`.

The problem is that `property_list::has_property` depends on many type traits which are scattered around several files and property-specific header files like the one I'm adding specialize them. It is not immediately clear to me how to re-organize the code without having to split an implementation of properties for an extension into several files.

As of right now `virtual_functions.hpp` is relatively small, but implementing a `FIXME` about proper handling of `calls_indirectly` property will bring a couple hundreds lines of template-heavy code and then the impact will be bigger. I suppose the refactoring should happen somewhere in between.

**Review by uditagarwal97 — COMMENTED** [review]
The changes LGTM. However, I'd request an additional review from @againull

File: sycl/include/sycl/ext/intel/experimental/virtual_functions.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      // FIXME: we should handle Rest... here as well
+      __builtin_sycl_unique_stable_name(First);
+#else
+      "";
+#endif
+};
+
+template <typename T = void> struct void_or_T {
+  using type = T;
+};
```
Comment (againull) [inline sycl/include/sycl/ext/intel/experimental/virtual_functions.hpp:85]: This doesn't seem to be used anywhere, should it be removed? Or is it for future use?

**Review by againull — APPROVED** [review]
LGTM, just one minor comment.

File: sycl/include/sycl/ext/intel/experimental/virtual_functions.hpp (hunk 2 above)
Comment (AlexeySachkov) [reply sycl/include/sycl/ext/intel/experimental/virtual_functions.hpp]: It was probably a part of a protype I had for handling of more than one template argument for `calls_indirectly`. I've removed it for now in 1fd0b0bfa126962a1f2bbdcc10f3002c9370660e

**Comment by AlexeySachkov** [comment]
There were commits to this PR after it had been approved, but they only contained minor changes, such as move from `intel` to `oneapi` namespace and some tests cleanup because a dependent patch was merged. Otherwise, the code is still very much the same as it was at time of approval. Having that said, I will proceed with the merge. Feel free to leave feedback nonetheless (if any), I will apply it in a separate PR.

**Pull request merged**